kernel/thread.c

```diff
@@ -29,17 +29,60 @@ extern struct _static_thread_data _static_thread_data_list_start[];
 extern struct _static_thread_data _static_thread_data_list_end[];
 
 #ifdef CONFIG_USERSPACE
-/* Each thread gets assigned an index into a permission bitfield */
-static atomic_t thread_index;
+static int thread_count;
 
-static unsigned int thread_index_get(void)
+/*
+ * Fetch an unused thread ID. Returns -1 if all thread IDs are in use
+ */
+static int thread_index_get(void)
 {
-	unsigned int retval;
+	int key, pos = -1;
+
+	key = irq_lock();
+
+	if (thread_count == CONFIG_MAX_THREAD_BYTES * 8) {
+		/* We have run out of thread IDs! */
+		goto out;
+	}
+
+	/* find an unused bit in the kernel's bitfield of in-use thread IDs */
+	for (int i = 0; i < CONFIG_MAX_THREAD_BYTES; i++) {
+		int fs;
+
+		fs = find_lsb_set(_kernel.free_thread_ids[i]);
+		if (fs) {
+			/* find_lsb_set counts bit positions starting at 1 */
+			--fs;
+			_kernel.free_thread_ids[i] &= ~(1 << fs);
+			pos = fs + (i * 8);
+			break;
+		}
+	}
+
+	thread_count++;
+out:
+	irq_unlock(key);
 
-	retval = (int)atomic_inc(&thread_index);
-	__ASSERT(retval < 8 * CONFIG_MAX_THREAD_BYTES,
-		 "too many threads created, increase CONFIG_MAX_THREAD_BYTES");
-	return retval;
+	return pos;
+}
+
+static void thread_index_free(int id)
+{
+	int index, key;
+	u8_t bit;
+
+	if (id == -1) {
+		return;
+	}
+
+	key = irq_lock();
+
+	thread_count--;
+	index = id >> 3;
```

Comment:
You can divide by 8 and mod with 8 here: these are constants, and the compiler will perform strength reduction at the lowest optimization levels.   Use shifts when you mean shifts, to maintain symmetry with the previous, related, function.

Reply:
done